ortb/clone_test.go

```diff
@@ -536,6 +536,326 @@ func TestCloneUser(t *testing.T) {
+			IFA:            "IFA",
+			DIDSHA1:        "DIDSHA1",
+			DIDMD5:         "DIDMD5",
+			DPIDSHA1:       "DPIDSHA1",
+			DPIDMD5:        "DPIDMD5",
+			MACSHA1:        "MACSHA1",
+			MACMD5:         "MACMD5",
+			Ext:            json.RawMessage(`{"anyField":1}`),
+		}
+		result := CloneDevice(given)
+		assert.Equal(t, given, result, "equality")
+		assert.NotSame(t, given, result, "pointer")
+		assert.NotSame(t, given.Geo, result.Geo, "geo")
+		assert.NotSame(t, given.Geo.Ext, result.Geo.Ext, "geo-ext")
+		assert.NotSame(t, given.DNT, result.DNT, "dnt")
+		assert.NotSame(t, given.Lmt, result.Lmt, "lmt")
+		assert.NotSame(t, given.SUA, result.SUA, "sua")
+		assert.NotSame(t, given.ConnectionType, result.ConnectionType, "connectionType")
+		assert.NotSame(t, given.Ext, result.Ext, "ext")
+	})
+
+	t.Run("assumptions", func(t *testing.T) {
+		assert.ElementsMatch(t, discoverPointerFields(reflect.TypeOf(openrtb2.Device{})),
+			[]string{
+				"Geo",
+				"DNT",
+				"Lmt",
+				"SUA",
+				"ConnectionType",
+				"Ext",
+			})
+	})
+}
+
+func TestCloneInt8Pointer(t *testing.T) {
+
+	t.Run("nil", func(t *testing.T) {
+		result := CloneInt8Pointer(nil)
+		assert.Nil(t, result)
+	})
+
+	t.Run("empty", func(t *testing.T) {
+		var given *int8
+		result := CloneInt8Pointer(given)
+		assert.Nil(t, result)
+	})
+
+	t.Run("populated", func(t *testing.T) {
+		var given *int8
+		var n int8
+		n = 1
+		given = &n
+		result := CloneInt8Pointer(given)
+		assert.Equal(t, given, result, "equality")
+		assert.NotSame(t, given, result, "pointer")
+	})
+}
+
+func TestCloneBrandVersionSlice(t *testing.T) {
+	t.Run("nil", func(t *testing.T) {
+		result := CloneBrandVersionSlice(nil)
+		assert.Nil(t, result)
+	})
+
+	t.Run("empty", func(t *testing.T) {
+		given := []openrtb2.BrandVersion{}
+		result := CloneBrandVersionSlice(given)
+		assert.Empty(t, result)
+		assert.NotSame(t, given, result)
+	})
+
+	t.Run("one", func(t *testing.T) {
+		given := []openrtb2.BrandVersion{
+			{Brand: "1", Version: []string{"s1", "s2"}, Ext: json.RawMessage(`{"anyField":1}`)},
+		}
+		result := CloneBrandVersionSlice(given)
+		assert.Equal(t, given, result, "equality")
+		assert.NotSame(t, given[0], result[0], "item-pointer")
+		assert.NotSame(t, given[0].Ext, result[0].Ext, "item-pointer-ext")
+	})
+
+	t.Run("many", func(t *testing.T) {
+		given := []openrtb2.BrandVersion{
+			{Brand: "1", Version: []string{"s1", "s2"}, Ext: json.RawMessage(`{"anyField":1}`)},
+			{Brand: "2", Version: []string{"s3", "s4"}, Ext: json.RawMessage(`{"anyField":1}`)},
+			{Brand: "3", Version: []string{"s5", "s6"}, Ext: json.RawMessage(`{"anyField":1}`)},
+		}
+		result := CloneBrandVersionSlice(given)
+		assert.Equal(t, given, result, "equality")
+		assert.NotSame(t, given[0], result[0], "item0-pointer")
+		assert.NotSame(t, given[0].Ext, result[0].Ext, "item0-pointer-ext")
+		assert.NotSame(t, given[1], result[1], "item1-pointer")
+		assert.NotSame(t, given[1].Ext, result[1].Ext, "item1-pointer-ext")
```

Comment:
The third brand version is not being asserted as different. We should add:
```
assert.NotSame(t, given[2], result[2], "item2-pointer")
assert.NotSame(t, given[2].Ext, result[2].Ext, "item2-pointer-ext")
```

Reply:
Added. HP comparison added too.